File: 인구이동.cpp
// https://www.acmicpc.net/problem/16234

#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#include <cstring>
#define MAXN 50
using namespace std;
typedef unsigned long long ull;
typedef long long ll;

int n,l,r;
int A[MAXN+1][MAXN+1];
int visited[MAXN+1][MAXN+1];

struct mov
{
    /* data */
    int x;
    int y;
};

// mov path[MAXN+1];

int dx[] = {-1,1,0,0};
int dy[] = {0,0,-1,1};

void input() {
    // freopen("in.txt","r",stdin);
    cin>>n>>l>>r;
    for(int i=0;i<n;i++){
        for(int j=0;j<n;j++){
            cin>>A[i][j];
        }
    }
}


int sol(int xx, int yy) {

    queue<mov>q;
    q.push({xx,yy});

    int sum = A[xx][yy];
    int cnt = 1;

    visited[xx][yy] = 1;
    
    vector<pair<int, int>> cand;
    cand.push_back(make_pair(xx, yy));

    while(!q.empty()){
        mov qv = q.front();q.pop();

        for(int i=0;i<4;i++){
            int nx = qv.x+dx[i];
            int ny = qv.y+dy[i];

            
            
            if(nx<0||ny<0||nx>=n||ny>=n) continue;
            if(visited[nx][ny] != -1) continue;

            int isOpen = abs(A[qv.x][qv.y]-A[nx][ny]);

            if(isOpen < l || isOpen > r ) continue;

            visited[nx][ny] = 1;
            sum += A[nx][ny];
            cnt += 1;
            q.push({nx,ny});

            cand.push_back(make_pair(nx, ny));

        }


    }

    int calc = sum / cnt;
    for(auto c: cand) {
        A[c.first][c.second] = calc;
        visited[c.first][c.second] = 2;
    }
    
    return cnt != 1;
}

int main() {
    ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);
    input();

    int ans=0;
    int ansCnt = 0;
    while(1){
        ansCnt++;
        memset(visited,-1,sizeof(visited));
        bool flag = false;
        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                if(visited[i][j] != -1) continue;
                flag |= sol(i,j);
                
            }   
        }
        
        if(!flag){
            cout<<ans;
            break;
        } else{
            ans += 1;
        }
    }
    
    
    
}

File: 인구이동.py
# https://www.acmicpc.net/problem/16234
import sys
from collections import deque
si = sys.stdin.readline
mis = lambda: map(int, si().split())

dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

def bfs(x, y, k):
    visited[x][y] = k
    
    q = deque([(x, y)])

    cand = [(x, y)]
    cnt = 1
    SUM = a[x][y]
    while q:
        x, y = q.popleft()
        
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if not (0 <= nx < n and 0 <= ny < n): continue
            if visited[nx][ny] != 0: continue

            d = abs(a[x][y] - a[nx][ny])
            if not (l <= d <= r): continue
            
            visited[nx][ny] = k
            q.append((nx, ny))
            cand.append((nx, ny))
            cnt += 1
            SUM += a[nx][ny]
    
    calc = SUM // cnt
    for x, y in cand:
        a[x][y] = calc
    
    return cnt != 1


if __name__ == '__main__':
    n, l, r = mis()
    a = [list(mis()) for _ in range(n)]

    ans = 0
    while True:
        visited = [[0] * n for _ in range(n)]
        
        k = 1
        flag = False
        for i in range(n):
            for j in range(n):
                if visited[i][j] != 0: continue
                flag |= bfs(i, j, k)
                k += 1
        
        if flag:
            ans += 1
        else:
            break
    print(ans)